proposals/0000-unicode-for-string-processing.md

```diff
@@ -0,0 +1,1321 @@
+
+With `RegexBuilder`'s `CharacterClass` type, you can use built-in character classes with ranges and groups of characters. For example, to parse a valid octodecimal number, you could define a custom character class that combines `.digit` with a range of characters.
+
+```swift
+let octoDecimalRegex: Regex<(Substring, Int?)> = Regex {
+    let charClass = CharacterClass(.digit, "a"..."h").ignoresCase()
+    Capture {
+      OneOrMore(charClass)
+    } transform: { Int($0, radix: 18) }
+}
+```
+
+The full `CharacterClass` API is as follows:
+
+```swift
+/// A class of characters that match in a regex.
+///
+/// A character class can represent individual characters, a group of
+/// characters, the set of character that match some set of criteria, or
+/// a set algebraic combination of all of the above.
+public struct CharacterClass: RegexComponent {
+  public var regex: Regex<Substring> { get }
+
+  /// A character class that matches any character that does not match this
+  /// character class.
+  public var inverted: CharacterClass { get }
+}
+
+// MARK: Built-in character classes
+
+extension RegexComponent where Self == CharacterClass {
+  /// A character class that matches any element.
+  ///
+  /// This character class is unaffected by the `dotMatchesNewlines()` method.
+  /// To match any character that isn't a newline, see
+  /// ``CharacterClass.anyNonNewline``.
+  ///
+  /// This character class is equivalent to the regex syntax "dot"
+  /// metacharacter in single-line mode: `(?s:.)`.
+  public static var any: CharacterClass { get }
+
+  /// A character class that matches any element that isn't a newline.
+  ///
+  /// This character class is unaffected by the `dotMatchesNewlines()` method.
+  /// To match any character, including newlines, see ``CharacterClass.any``.
+  ///
+  /// This character class is equivalent to the regex syntax "dot"
+  /// metacharacter with single-line mode disabled: `(?-s:.)`.
+  public static var anyNonNewline: CharacterClass { get }
+
+  /// A character class that matches any single `Character`, or extended
+  /// grapheme cluster, regardless of the current semantic level.
+  ///
+  /// This character class is equivalent to `\X` in regex syntax.
+  public static var anyGraphemeCluster: CharacterClass { get }
+
+  /// A character class that matches any digit.
+  ///
+  /// This character class is equivalent to `\d` in regex syntax.
+  public static var digit: CharacterClass { get }
+  
+  /// A character class that matches any hexadecimal digit.
+  public static var hexDigit: CharacterClass { get }
+
+  /// A character class that matches any element that is a "word character".
+  ///
+  /// This character class is equivalent to `\w` in regex syntax.
+  public static var word: CharacterClass { get }
+
+  /// A character class that matches any element that is classified as
+  /// whitespace.
+  ///
+  /// This character class is equivalent to `\s` in regex syntax.
```

Comment:
Is this affected by ascii only? Yes/no? Should there be an additional one that is or isn't affected?